cpp/test/core/mdspan_copy.cpp

```diff
@@ -0,0 +1,275 @@
+/*
+ * Copyright (c) 2023, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "../test_utils.h"
+#include <cstdint>
+#include <gtest/gtest.h>
+#include <raft/core/copy.hpp>
+#ifndef RAFT_DISABLE_CUDA
+#include <raft/core/device_mdarray.hpp>
+#include <raft/core/device_resources.hpp>
+#endif
+#include <raft/core/host_mdarray.hpp>
+#include <raft/core/resources.hpp>
+
+namespace raft {
+TEST(MDSpanCopy, Mdspan1DHostHost)
```

Comment:
Can you add another test here where you explicitly construct `detail::mdspan_copyable` and for all possible input type combinations, you ensure the right internal booleans that denote which policy to use for copying are being set?

Reply:
I'd prefer not to include implementation details in unit tests. Ideally, the tests should ensure correctness for every scenario that can be handled by the public API, and if we ultimately are taking the wrong path (but still giving correct results), then that's a benchmarking question. That is to say, so long as we have the correct results, all we care about is whether the copy is as fast as it should be and not using up system resources unnecessarily.

Reply:
I think I disagree here that arriving at the correct result using the wrong path is a correct test. I would very much like to see a unit test that checks the internals of `mdspan_copyable` as it's not only a way to ensure type correctness, but also helps any person who modifies this code figure out that they did not change an internal that leads them to take the wrong path by mistake and spend time on benchmarks figuring out why perf suddenly degraded.

A test with clear compile-time asserts will make changes easier to iterate on for any future dev working on this.